File: src/matlab/weightsQ.m
% SPDX-License-Identifier: GPL-3.0-only
% 
% Copyright 2008-2024 San Diego State University (SDSU) and Contributors 
%
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, version 3.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% LICENSE file or on the web GNU General Public License 
% <https://www.gnu.org/licenses/> for more details.
%
% ------------------------------------------------------------------------

function Q = weightsQ(k, m, dx)
% Returns the m+2 weights of Q
%
% Parameters:
%                k : Order of accuracy
%                m : Number of cells
%               dx : Step size

    D = div(k, m, dx);
    
    b = [-1; zeros(m-1, 1); 1]; % RHS
    
    Q = [1; D(2:end-1, :)'\b; 1];
end

File: src/cpp/weightsQ.cpp
/*
* SPDX-License-Identifier: GPL-3.0-or-later
* © 2008-2024 San Diego State University Research Foundation (SDSURF).
* See LICENSE file or https://www.gnu.org/licenses/gpl-3.0.html for details. 
*/

/*
 * @file weights.cpp
 * 
 * @brief Mimetic Differences Q Weights
 * 
 * @date 2025/10/14
 * 
 */

#include "utils.h"
#include "divergence.h"
#include "gradient.h"
#include "weightsQ.h"

WeightsQ::WeightsQ(u16 k, u32 m, Real dx): sp_vec(m + 1) 
{
  Divergence D(k, m, dx);
  D.shed_row(0);
  D.shed_row(m);

  vec b(m+1);
  b.at(0) = -1.0;
  b.at(m) = 1.0;

  sp_mat Dtranspose = D.t();
  sp_vec Q_prime = Utils::spsolve_eigenQR(Dtranspose, b);
  sp_vec Q(m+2);
  Q.at(0) = 1.0;
  Q.at(m+1) = 1.0;
  for (int i = 1; i < m+1; ++i) {
    Q.at(i) = Q_prime.at(i-1);
  }

  *this = Q;

}

